src/input.rs

```diff
@@ -623,6 +644,123 @@ where
 {
 }
 
+/// An input wrapper that returns a custom span, with the user-defined context
+/// contained in the Span::Context. See [`Input::with_span_context`].
+#[derive(Copy, Clone)]
+pub struct WithSpanContext<S: Span, I> {
+    input: I,
+    context: S::Context,
+    phantom: PhantomData<S>,
+}
```

Comment:
Re-pushed this branch changing `WithContext` to be this behaviour. 

Also commented on #447 re: `map_span`.

If you want to have both, merge this whenever (if you're happy with the other changes). If you're only wanting one, then I have some time next weekend to dabble in what an implementation of `map_span` might look like, and we could regroup after that.

Reply:
Actually update @zesterer:

Played a little with a map_span impl. It seems hard:

We'd need some `MappedSpan` input wrapper containing the closure. As I'd need the closure to capture variables (e.g. the filename), it can't be a `fn` pointer. So it would have to be something like `dyn Fn(I::Span) -> S`, which now means the `MappedSpan` couldn't be `Copy` (or even `Clone` without boxing), and it feels like a snowball growing as I'm sure that's required elsewhere.

The `WithContext` implementation in this PR works well for me because I can also have the `Span::Context` borrow from the input, and it seems hard to wrangle the closure to work the same way without a lot of indirection (or at all (?)).

So, I think this PR is the way to go for now 😅 but if you have any good ideas on the `map_span` implementation let me know and I can re-dabble!

Reply:
Why not simply bind `MappedSpan` with the function type?

```rs
struct MappedSpan<I, F> {
    input: I,
    mapper: F,
}
```

Reply:
That does work, but ultimately when using it this line would need to specify a concrete type: https://github.com/bglw/chumsky/blob/ea3036f9885562d55f6416b96874f6bd7d177d8c/src/lib.rs#L2689

Instead of `WithContext<Span, &'a str>` we would need something like:

```rust
MappedSpan<Span<'a>, &'a str, fn(SimpleSpan<usize>) -> Span<'a>>
```

But a closure won't coerce here, and it'd have to be `dyn Fn . . .` to use the trait, which then requires some kind of boxing, and seems hard on the usage-side to manage the lifetimes of anything borrowed

Reply:
Well, you can always make the parser generic over the input type. I recommend doing this anyway for non-trivial input types (of which MappedSpan would surely be an example).

Reply:
Alright I've pushed a commit containing that as well. Helper:

https://github.com/zesterer/chumsky/blob/6449929bb9908437d4d026557c4adec81555d63f/src/input.rs#L120-L135

and this is the best usage I could come up with:

bounds, making the parser generic over that function specifically:
https://github.com/zesterer/chumsky/blob/6449929bb9908437d4d026557c4adec81555d63f/src/lib.rs#L2738-L2739

usage (with a borrowed value):
https://github.com/zesterer/chumsky/blob/6449929bb9908437d4d026557c4adec81555d63f/src/lib.rs#L2759-L2776